Userland/Libraries/LibJS/AST.cpp

```diff
@@ -888,10 +836,17 @@ Value ForOfStatement::execute(Interpreter& interpreter, GlobalObject& global_obj
         interpreter.vm().running_execution_context().lexical_environment = old_environment;
     });
 
-    (void)TRY_OR_DISCARD(get_iterator_values(global_object, rhs_result, [&](Value value) -> Optional<Completion> {
+    auto last_value = js_undefined();
+
+    (void)TRY(get_iterator_values(global_object, rhs_result, [&](Value value) -> Optional<Completion> {
         TRY(for_of_head_state.execute_head(interpreter, global_object, value));
-        last_value = m_body->execute(interpreter, global_object).value_or(last_value);
+        auto result = m_body->execute(interpreter, global_object);
         interpreter.vm().running_execution_context().lexical_environment = old_environment;
+        // NOTE: We need to do this out of order as we can't directly return from here as we're
+        // supposed to, since this is the get_iterator_values() callback. Instead, we have to rely
+        // on updating the outer last_value.
+        if (result.value().has_value())
+            last_value = *result.value();
```

Comment:
Kind of the same question as with for in can we not just check the result?

Reply:
Same answer, it's temporary :^)